### PR TITLE
XEP-0388: Rework whole spec, namespace bump

### PR DESCRIPTION
**Pull request by tmolitor-stud-tu**
This is a collaboration between @mwild1 and me. Most parts have already been discussed with the original author @dwd, too.

In short, this adds:
- Namespace bump
- Make SCRAM hashes (and other SASL mechanism families) upgradable
- Add dependency on XEP-0440 for channel-binding agility
- Add new `<inline/>` element as extension point to inline stream resumption and bind2 (and possibly others)
- Clarify of `<continue/>` and tasks in general
- Update security considerations and business rules
- Add optional `<user-agent/>` element

### CODE REVIEW

**Comment by horazont** [comment]
@dwd Ack please?

**Comment by tmolitor-stud-tu** [comment]
For reference, here is the rendered version: https://dyn.eightysoft.de/final/xep-0388.html

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:304]: I designed tasks to be syntactically aligned with SASL mechanisms, partly in the hope that other protocols might adopt them, and partly because it allowed for an easily pluggable system on the endpoints (like SASL mechanisms themselves), and partly because it meant that there are only SASSL2 top level elements in flow during the entirety of the SASL2 exchange, making a few bits of housekeeping simpler.

The fact you've elected to switch to XML is in itself fine, but it does mean that the exchange now introduces arbitrary elements, and the SASL2 code itself needs to be more aware of what tasks may be in flight.

I'm not averse to using XML here for tasks; I don't honestly have much hope for (for example) LDAP or IMAP using them. But I do think that means you'll need other framing to handle the now-discarded properties.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:328]: Every time you impose a restriction on the string syntax of an opaque identifier, a layer-violation fairy dies.

Less prosaically, either a client knows how to upgrade to a mechanism or else it doesn't, and either way the "UPGR-" doesn't really help.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:354]: See above comments on switching from base64 to XML. Even a <task xmlns='urn:xmpp:sasl:2'/> wrapper around these would simplify a lot of things.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:392]: See above - all worthy things, but baking a particular SASL mechanism into the SASL profile seems entirely counter to the principles of SASL.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:413]: I hate restating things from other specifications, but I accept this is a personal quibble. And in any case, you can defer to the RFC for the weakening of the SHOULD NOT:

As explained in ... servers and clients SHOULD NOT support SASL PLAIN.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:414]: No it doesn't provide or allow for that - the only hint is provided by facilities in RFC 6120 itself. "This protocol makes use of the hitn contained in the stream "from" attribute ..."?

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:415]: I'd actually specify this as a mitigation to the first three points.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:416]: Again, we're somewhat into MTI territory .

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:417]: I would make this a SHOULD NOT. I think you list enough explanation and potential mitigations to cover it.

File: xep-0388.xml
Comment (dwd) [inline xep-0388.xml:181]: OK, I could be persuaded that <user-agent/> itself, with the id, is a good idea. I can see the utility in it as a useful baseline.

But the way it's written, <software/> and <device/> seem to be both mandatory, and I think that's definitely wrong.

**Review by dwd — COMMENTED** [review]
There's a *lot* of changes to digest here, beyond even those we discussed (and I reserve my right to change my mind even when we did discuss!).

I don't think this is sensible to publish yet, and certainly not without extensible discussion on the list.

**Comment by Flowdalic** [comment]
> For reference, here is the rendered version: https://dyn.eightysoft.de/final/xep-0388.html

It would probably help to have a diff from the current version. Especially for an upcoming list disccusion. I usually [create HTML diffs for XEPs](https://github.com/Flowdalic/xeps/blob/e31cf7c009f3a30558bda947d787e61e40ecfe39/xep-template/Makefile#L66-L70) using [htmldiff](https://github.com/cygri/htmldiff).

**Comment by mwild1** [comment]
I generally agree with @dwd's points. The main thing I'll defend is having `<user-agent>` in here.

Honestly, I spent the past 5 years somehow thinking SASL2 *already* had a client identifier in it. I guess I made that up somewhere along the way, but I do think it makes sense. It's not the end of the world if it's in an separate XEP, but it's a solution to so many problems, and necessary for some SASL mechanisms (e.g. `HT-*`) because they don't provide any other key via which the server can determine which credentials it should be comparing against (maybe that's something to fix in those mechanisms, but there are other use cases too, including just debugging and diagnostic purposes).

As for XEP-0198 integration, the intent is not to define it in this spec, merely to say it's possible. It's covered by PR #1215 updating XEP-0198 itself. I prefer that PR over XEP-0397, as it reuses more of the existing protocol and fulfils more use cases without involving any extra stuff (e.g. token auth mechanisms).

Anyway, this should all be on the mailing list. If someone sparks a thread, I'll be happy to ^C^V everything there. If nobody does, I'll start one soon.

File: xep-0388.xml
Comment (iNPUTmice) [reply xep-0388.xml:415]: Obligatory comment that they are other methods to secure a stream than TLS.

File: xep-0388.xml
Comment (iNPUTmice) [reply xep-0388.xml:38]: To be clear `<inline/>` is merely a place where SASL 2 extensions can announce themselves (instead of being direct stream features).
But I don’t have any strong feelings on whether stream feature or inline stream feature is the cleaner approach. One signals that it depends on SASL2 the other one would allow use to use Bind 2 over Sasl 3.

**Comment by tmolitor-stud-tu** [comment]
> There's a _lot_ of changes to digest here, beyond even those we discussed (and I reserve my right to change my mind even when we did discuss!).

Sure! I did not intend to imply that our discussionwas something like a pre-approval. I just wanted to note that you are informed that I was working on this XEP. Of course you can change your mind any time.

> I don't think this is sensible to publish yet, and certainly not without extensible discussion on the list.

I think the way forward if to address some of your issues we don't need to discuss about and leave the rest for an open discussion on the list. Does this seem sensible to you?

_Quick warning: I'm really short at time currently. Several days may pass until I'm able to answer._

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:35]: "dependency" is meant as "refers to". I'll change that to say `Add reference to &xep0440; and &rfc5802;`

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:38]: SASL3 could well define an extension point, too. I don't think it's neccessary to remove the `<inline/>` just because of an hypothetical SASL3.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:34]: Well, yes and no. The upgrade tasks themselves probably still belong into SASL2. SCRAM updates on the other hand could be in an extra XEP as well.

I merely added them to SASL2 because it was easier and the SCRAM update part was relatively small compared to the rest of the SASL2 XEP.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:125]: The "MUST implement channel-binding" elsewhere could well be a bug and left over from an earlier iteration of the XEP.

A separate XEP for all the MTI stuff is a good idea, but I think the "if channel-binding, XEP-0440 is a MUST" should still be in SASL2.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:255]: That's a bug.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:273]: Oh, I didn't know this.
I thought all these different error conditions defined in SCRAM could be of real value.

I'll remove this sentence.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:284]: Hmm, I'll clarify this.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:328]: I wanted to make sure these tasks could eventually be used as SASL mechanism without introducing name clashes. I used `UPGR-` instead of `UPGRADE-` to keep the name below the 20 chars limit of SASL.
(`UPGR-` and `-PLUS` have the same size, and `-PLUS` won't be used for upgrades.

But I don't have a strong opinion about that.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:392]: Well, as I said above I thought defining the SCRAM upgrade task was small addition and didn't want to introduce yet another new XEP, but of course we could move that to a new XEP.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:413]: I've changed that to defer to the RFC :)

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:414]: changed that, too, thanks!

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:415]: @iNPUTmice RFC 6120 contains the same wording for SASL1.

That's the reason I wanted to state the same for SASL2 as well.

@dwd why would mandatory TLS be a mitigation for these 3 points?

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:416]: ah sorry, I can change that to a SHOULD. That's the MTI for channel-binding you are referring above, right?

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:417]: hmmm, any thoughts @mwild1 ?

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:181]: see @mwild1's response on list

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:39]: @mwild1 that one is for you, I guess.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml]: You can mark this as resolved if it was the MTI you referred to above.

File: xep-0388.xml
Comment (mwild1) [reply xep-0388.xml:416]: I've no strong feelings, and don't object to this change. I'm curious why you would relax a 'MUST NOT' that relates to a security concern, but sure... maybe there are closed trusted environments that want to do TLS 0-RTT without negotiation or replay protection. I would likewise hope that the explanation is enough to tell developers not to do this under any scenario where there might be a risk.

FWIW the TLS spec itself [says](https://www.rfc-editor.org/rfc/rfc8446#section-8):

> Therefore, in normal operation, clients will not know which, if any, of these mechanisms servers actually implement and hence MUST only send early data which they deem safe to be replayed.

File: xep-0388.xml
Comment (mwild1) [reply xep-0388.xml:39]: My response to this is in [this comment](https://github.com/xsf/xeps/pull/1214#issuecomment-1279753150) and [this mailing list post](https://mail.jabber.org/pipermail/standards/2022-October/038998.html).

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:391]: I'm moving the SCRAM part of upgrade tasks into a new ProtoXEP.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:34]: I've split out the SCRAM upgrade task specification into a new ProtoXEP.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:38]: I think this is already discussed on-list. I'll mark that as resolved here for now.

**Comment by tmolitor-stud-tu** [comment]
I marked everything addressed in my last commits as "solved" to redirect the focus to unresolved things.
Feel free to "unresolve" things if you see the need for additional changes and/or start a discussion about this on-list.

**Comment by tmolitor-stud-tu** [comment]
lgtm

current rendered version: https://dyn.eightysoft.de/final/xep-0388.html
current diff: https://dyn.eightysoft.de/final/xep-0388-diff.html

@mwild1 lgtm, too?

File: xep-0388.xml
Comment (dwd) [reply xep-0388.xml:339]: You're presumably implying that if TLS is negotiated, the risk due to the first three points is mitigated. I'd state that explicitly.

File: xep-0388.xml
Comment (dwd) [reply xep-0388.xml:340]: It's more that if you're using a multi-trip mechanism like SCRAM, then 0-RTT is fine for sending the SCRAM initial response, and it's no more or less secure. An attacker can at best repeat the client nonce, but that's hardly an issue. The problem is more likely to be that channel binding is trickier.

**Comment by tmolitor-stud-tu** [comment]
> It's more that if you're using a multi-trip mechanism like SCRAM, then 0-RTT is fine for sending the SCRAM initial response, and it's no more or less secure. An attacker can at best repeat the client nonce, but that's hardly an issue. The problem is more likely to be that channel binding is trickier.

We just did not want to go down this rabbit hole and rather leave the 0-RTT stuff to be weakened and defined in more detail by other/future XEPs:
It might work for SCRAM, but there might be other/future SASL mechanisms that would be more vulnerable if 0-RTT was used. Going the "default deny" route rather than the "default allow" seems like a sensible thing to do here.

**Comment by tmolitor-stud-tu** [comment]
> You're presumably implying that if TLS is negotiated, the risk due to the first three points is mitigated. I'd state that explicitly.

I still don't understand what "3 points" you are referring to.

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml:117]: Adding a sentence that TLS SHOULD be in place when specifying the from-attribute is a good idea.

The from-attribute is a MUST, because SASL2 is designed to not work properly without it for most authentication methods: the server can not specify in it's stream features which methods it supports unless it knows which user wants to authenticate.

That's okay only, if the server only offers mechanisms like SASL EXTERNAL etc.

That means the `if the "from" attribute is missing then this may incur an additional round-trip` is an explanation why the protocol was designed the way it is rather than an explanation for the MUST.

Comment (tmolitor-stud-tu) [reply xep-0388.xml:117]: Thinking about this again, I think this whole paragraph should be moved to the upgrade tasks ProtoXEP.
Without upgrade tasks (and the resulting dependence of the mechanism list on the user in the from attribute), giving the from attribute won't even save an additional round-trip: SASL2 behaves exactly like SASL1 here.

**Comment by dwd** [comment]
> > It's more that if you're using a multi-trip mechanism like SCRAM, then 0-RTT is fine for sending the SCRAM initial response, and it's no more or less secure. An attacker can at best repeat the client nonce, but that's hardly an issue. The problem is more likely to be that channel binding is trickier.
> 
> We just did not want to go down this rabbit hole and rather leave the 0-RTT stuff to be weakened and defined in more detail by other/future XEPs: It might work for SCRAM, but there might be other/future SASL mechanisms that would be more vulnerable if 0-RTT was used. Going the "default deny" route rather than the "default allow" seems like a sensible thing to do here.

You're not doing "default deny", you're doing "deny".

**Comment by dwd** [comment]
> > You're presumably implying that if TLS is negotiated, the risk due to the first three points is mitigated. I'd state that explicitly.
> 
> I still don't understand what "3 points" you are referring to.

The first three points made in this section, about additional exposure and so on.

File: xep-0388.xml
Comment (dwd) [reply xep-0388.xml]: But... All you're doing is raising the SHOULD in RFC6120 §4.7.1 to MUST. If you restate everything else people will think any omissions are intentional.

Comment (tmolitor-stud-tu) [reply xep-0388.xml]: Done

**Comment by tmolitor-stud-tu** [comment]
> > > It's more that if you're using a multi-trip mechanism like SCRAM, then 0-RTT is fine for sending the SCRAM initial response, and it's no more or less secure. An attacker can at best repeat the client nonce, but that's hardly an issue. The problem is more likely to be that channel binding is trickier.
> > 
> > 
> > We just did not want to go down this rabbit hole and rather leave the 0-RTT stuff to be weakened and defined in more detail by other/future XEPs: It might work for SCRAM, but there might be other/future SASL mechanisms that would be more vulnerable if 0-RTT was used. Going the "default deny" route rather than the "default allow" seems like a sensible thing to do here.
> 
> You're not doing "default deny", you're doing "deny".

But that "deny" could be overwritten by a later xep, no?

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml]: I've moved that MUST to the protoXEP and rephrased it. It should be clearer now: https://dyn.eightysoft.de/final/xep-scram-upgrade.html#protocol
Or do you think it's still not clear?

**Comment by tmolitor-stud-tu** [comment]
> > > You're presumably implying that if TLS is negotiated, the risk due to the first three points is mitigated. I'd state that explicitly.
> > 
> > 
> > I still don't understand what "3 points" you are referring to.
> 
> The first three points made in this section, about additional exposure and so on.

Ah, I understand.
Fixed here: https://github.com/xsf/xeps/pull/1214/commits/18fb29343be4c8ada47bb23450d2c08ce2e46afe

File: xep-0388.xml
Comment (tmolitor-stud-tu) [reply xep-0388.xml]: @horazont over here

**Review by dwd — APPROVED** [review]

**Comment by Neustradamus** [comment]
To follow

**Comment by tmolitor-stud-tu** [comment]
@Kev this isn't "needs author" anymore, Dave already confirmed this on standards@

@dwd can you confirm this here again?

**Comment by tmolitor-stud-tu** [comment]
@dwd I'm sorry, I can't find your mail at standards@ confirming this can be merged. Am i hallucinating?

**Comment by tmolitor-stud-tu** [comment]
@Kev This is not NeedsAuthor anymore, but ReadyToMerge!

See here: https://mail.jabber.org/pipermail/standards/2023-January/039133.html

**Comment by guusdk** [comment]
@tmolitor-stud-tu I believe you can make the XEP validation happy if you reverse the order of lines 15 and 16. `<sig>` should go before `<approver>`, not the other way around.

**Comment by tmolitor-stud-tu** [comment]
> Update: I was allowed to and have pushed a commit with this change to your branch.
Perfect, thanks!

**Comment by guusdk** [comment]
XEP state is currently 'deferred'. Changes are _not_ purely editorial. As per https://github.com/xsf/xeps/blob/master/docs/TRIAGING.md this requires a move to 'experimental' state.

**Comment by guusdk** [comment]
In an attempt to move this forward (and after consultation in editors@) I've added a commit that changes the status from the XEP from Deferred to Experimental.

**Comment by guusdk** [comment]
> not doing what I asked

I blame poor instructions :)

Should be fixed now.

**Pull request merged**
